packages/online-editor/src/workspace/components/WorkspaceStatusIndicator.tsx

```diff
@@ -28,61 +28,214 @@ import { SecurityIcon } from "@patternfly/react-icons/dist/js/icons/security-ico
+  workspaces: WorkspacesContextType;
+  workspaceGitStatusPromise?: PromiseState<WorkspaceGitStatusType>;
+  workspaceFile?: WorkspaceFile;
+  onStatusChanged?: (status?: "modified" | "deleted" | "synced") => void;
+  children?: JSX.Element;
+  isHoverable?: boolean;
+}) {
+  // We use this trick to prevent the icon from blinking while updating.
+  const prev = usePrevious(props.workspaceGitStatusPromise?.data);
+
+  const deletedFilesForWorkspace = React.useMemo(() => {
+    return listDeletedFiles({
+      workspaceDescriptor: props.workspaceDescriptor,
+      workspaceGitStatus: props.workspaceGitStatusPromise?.data,
+    });
+  }, [props.workspaceDescriptor, props.workspaceGitStatusPromise]);
+
+  const isThereDeletedFile = React.useMemo(() => {
+    const maybeWorkspaceFile = props.workspaceFile;
+    return (
+      (!maybeWorkspaceFile && deletedFilesForWorkspace.length) ||
+      (maybeWorkspaceFile &&
+        deletedFilesForWorkspace.some((deleted) => deleted.relativePath === maybeWorkspaceFile.relativePath))
+    );
+  }, [deletedFilesForWorkspace, props.workspaceFile]);
+
+  const localChangesStatus = React.useMemo(() => {
+    const status = resolveGitLocalChangesStatus({
+      workspaceGitStatus: props.workspaceGitStatusPromise?.data ?? prev,
+      file: props.workspaceFile,
+    });
+    if (props.onStatusChanged) {
+      props.onStatusChanged(
+        switchExpression(status, {
+          pending: isThereDeletedFile ? "deleted" : "modified",
+          synced: "synced",
+          default: undefined,
+        })
+      );
+    }
+    return status;
+  }, [isThereDeletedFile, prev, props]);
+
+  const indicatorTooltips = React.useMemo(() => {
+    if (!props.workspaceGitStatusPromise?.data?.hasLocalChanges) {
+      return [];
+    }
+    const tooltips: JSX.Element[] = [];
+    const modifiedTooltip = (
+      <Tooltip content={"Modified file in the Workspace."} position={"bottom"} key="modified">
+        <small>
+          <i>M</i>
+        </small>
+      </Tooltip>
+    );
+    const deletedTooltip = (
+      <Tooltip content={"Deleted file from the Workspace."} position={"bottom"} key="deleted">
+        <small>
+          <i>D</i>
+        </small>
+      </Tooltip>
+    );
+    const addedTooltip = (
+      <Tooltip content={"New file added into the Workspace."} position={"bottom"} key="added">
+        <small>
+          <i>A</i>
+        </small>
+      </Tooltip>
+    );
+    const tooltipForStageStatus = (stageStatus?: FileStageStatus) =>
+      switchExpression(stageStatus, {
+        added: addedTooltip,
+        modified: modifiedTooltip,
+        deleted: deletedTooltip,
+        default: <></>,
+      });
+    const maybeWorkspaceFile = props.workspaceFile;
+    if (maybeWorkspaceFile) {
+      tooltips.push(
+        tooltipForStageStatus(
+          props.workspaceGitStatusPromise?.data.fileStageStatuses.find(
+            (it) => it.path === maybeWorkspaceFile.relativePath
```

Comment:
If you don't know what to call a variable, you can resort to object destruction... 
```suggestion
            ({path}) => path === maybeWorkspaceFile.relativePath
```